cdc/log.cc

```diff
@@ -974,11 +976,12 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
         return make_ready_future<std::tuple<std::vector<mutation>, cdc::result_callback>>(std::make_tuple(std::move(mutations), result_callback{}));
     }
 
+    tracing::trace(tr_state, "CDC: Started generating mutations for log rows");
     mutations.reserve(2 * mutations.size());
 
     return do_with(std::move(mutations), service::query_state(service::client_state::for_internal_calls(), empty_service_permit()),
-            [this, timeout, i] (std::vector<mutation>& mutations, service::query_state& qs) {
-        return transform_mutations(mutations, 1, [this, &mutations, timeout, &qs] (int idx) {
+            [this, timeout, i, tr_state = std::move(tr_state)] (std::vector<mutation>& mutations, service::query_state& qs) {
+        return transform_mutations(mutations, 1, [this, &mutations, timeout, &qs, tr_state] (int idx) mutable {
```

Comment:
This won't work: "mutable" isn't going to cast away constantness. You need to instrument all lambdas in the call tree.

Reply:
Hmm, let's consider all `std::move(tr_state)`-s in this function (there are only two of them):

1. Capture at 983 moves out from the argument at 968 - OK, the argument isn't const,
1. Capture at 1005 moves out of capture at 984 - looks OK, the capturing lambda is mutable.

I tried verifying non-constness of the variables I was moving from by inserting `tr_state = nullptr` instructions one line before each of the `std::move`s, but the compilation failed.

It seems that when doing a simple capture by-copy (by placing only the name of the of the variable in the capture list, e.g. `[tr_state]`) in a mutable lambda, GCC does not remove constness from the type. I had to change the capture on the line above from `tr_state` to `tr_state = tr_state` to make `std::move`s from that variable work correctly.

Curiously, only GCC does that - in Clang and MSVC, variables captured that way in mutable lambdas seem to be non-const ([example on godbolt](https://godbolt.org/z/AWrzHK)).

Reply:
> Hmm, let's consider all `std::move(tr_state)`-s in this function (there are only two of them):
> 
>     1. Capture at 983 moves out from the argument at 968 - OK, the argument isn't const,
> 
>     2. Capture at 1005 moves out of capture at 984 - looks OK, the capturing lambda is mutable.

In order for tr_state to be not const you needed to make the lambda at 983 `mutable` .

I feel like there is some misunderstanding. Let me try to explain this in an example:

### Example 1 (bad)
```
some_movable_class var = ...;

// var is not const here
[var = std::move(var) ] {
    // var is const here
    [var = std::move(var)] () mutable { // var is going to be copied!
    }
}
```

### Example 2 (good)
```
some_movable_class var = ...;

// var is not const here
[var = std::move(var) ] () mutable{ // note the 'mutable' qualifier
    // var is not (!!!) const here
    [var = std::move(var)] () mutable { // var is moved
    }
}
```

`std::move()` is about r-values in general, where 'r' comes from "right".
It's important to make sure that the right operand in the assignment is not const.

> 
> 
> I tried verifying non-constness of the variables I was moving from by inserting `tr_state = nullptr` instructions one line before each of the `std::move`s, but the compilation failed.
> 
> It seems that when doing a simple capture by-copy (by placing only the name of the of the variable in the capture list, e.g. `[tr_state]`) in a mutable lambda, GCC does not remove constness from the type. I had to change the capture on the line above from `tr_state` to `tr_state = tr_state` to make `std::move`s from that variable work correctly.
> 
> Curiously, only GCC does that - in Clang and MSVC, variables captured that way in mutable lambdas seem to be non-const ([example on godbolt](https://godbolt.org/z/DkfkuR)).